data-prepper-plugins/user-agent-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/useragent/UserAgentProcessor.java

```diff
@@ -0,0 +1,115 @@
+
+import static org.opensearch.dataprepper.logging.DataPrepperMarkers.EVENT;
+
+@DataPrepperPlugin(name = "user_agent", pluginType = Processor.class, pluginConfigurationType = UserAgentProcessorConfig.class)
+public class UserAgentProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(UserAgentProcessor.class);
+    private final UserAgentProcessorConfig config;
+    private final Parser userAgentParser;
+
+    @DataPrepperPluginConstructor
+    public UserAgentProcessor(final PluginMetrics pluginMetrics, final UserAgentProcessorConfig config) {
+        super(pluginMetrics);
+        this.config = config;
+        this.userAgentParser = new CachingParser();
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            final String userAgentStr = event.get(config.getSource(), String.class);
+            if (Objects.isNull(userAgentStr)) {
+                continue;
+            }
+
+            try {
+                final Client clientInfo = this.userAgentParser.parse(userAgentStr);
+
+                final Map<String, Object> parsedUserAgent = getParsedUserAgent(clientInfo);
+                if (!config.getExcludeOriginal()) {
+                    parsedUserAgent.put("original", userAgentStr);
+                }
+                event.put(config.getTarget(), parsedUserAgent);
+            } catch (Exception e) {
+                LOG.error(EVENT, "An exception occurred when parsing user agent data from event [{}] with source key [{}]",
+                        event, config.getSource(), e);
+            }
+        }
+        return records;
+    }
+
+    @Override
+    public void prepareForShutdown() {
+    }
+
+    @Override
+    public boolean isReadyForShutdown() {
+        return true;
+    }
+
+    @Override
+    public void shutdown() {
+    }
+
+    private Map<String, Object> getParsedUserAgent(Client clientInfo) {
+        String version = getFullVersion(clientInfo.userAgent.major, clientInfo.userAgent.minor, clientInfo.userAgent.patch);
+
+        Map<String, Object> parsedUserAgent = new HashMap<>();
+        parsedUserAgent.put("name", clientInfo.userAgent.family);
+        parsedUserAgent.put("version", version);
+        parsedUserAgent.put("os", getParsedOS(clientInfo));
+        parsedUserAgent.put("device", getParsedDevice(clientInfo));
+
+        return parsedUserAgent;
+    }
+
+    private String getFullVersion(String major, String minor, String patch) {
+        if (Objects.isNull(major)) {
+            return "";
+        } else if (Objects.isNull(minor)) {
+            return major;
+        } else if (Objects.isNull(patch)) {
+            return String.format("%s.%s", major, minor);
+        }
+        return String.format("%s.%s.%s", major, minor, patch);
```

Comment:
`String.format` is quite slow. As this will be run repeatedly, let's use string addition.

I do think that testing with `String.format` is good. And I recommend it since it lets the tests retain the clarity that `String.format` provides.

Reply:
These String.format usages have been replaced.